### PR TITLE
Sni proxy protocol

### PR DESCRIPTION
**Pull request by tsloughter**
This PR adds support for v2 of proxy protocol. If the first line of a request is empty (because the first bytes are <<"\r\n">>) we know to expect v2 not v1. Information about the connection sent by the proxy protocol (like `sni_hostname`) is stored in a list `connection_info` in the `proxy_socket` record.

### CODE REVIEW

**Comment by ferd** [comment]
Do we want to put in the client code along with this PR somehow (https://gist.github.com/ferd/f0c172feee26b2fd9cd3) just to make it easier to work and test stuff in this repo when depending on it?

File: src/ranch_proxy_protocol.erl [hunk 1]
```diff
@@ -260,11 +329,91 @@ parse_proxy_protocol(<<"TCP", Proto:1/binary, _:1/binary, Info/binary>>) ->
...
             end
     end;
-parse_proxy_protocol(<<"UNKNOWN", _/binary>>) ->
+parse_proxy_protocol_v1(<<"UNKNOWN", _/binary>>) ->
     unknown_peer;
-parse_proxy_protocol(_) ->
+parse_proxy_protocol_v1(_) ->
+    not_proxy_protocol.
+
+%% first 4 bits are the version of the protocole, must be 2
```
Comment (evanmcc) [inline src/ranch_proxy_protocol.erl]: would prefer a diagram, but at the least there are numerous typos in this comment block that could be corrected.

File: src/ranch_proxy_protocol.erl [hunk 2]
```diff
@@ -48,6 +54,34 @@
 
 -define(DEFAULT_PROXY_TIMEOUT, config(proxy_protocol_timeout)).
 
+%% \r\n\r\n\0\r\nQUIT\n
+-define(HEADER, "\r\n\r\n\0\r\nQUIT\n").
```
Comment (evanmcc) [inline src/ranch_proxy_protocol.erl]: duplicative defines.

File: src/ranch_proxy_protocol.erl (hunk 1 above)
Comment (omarkj) [inline src/ranch_proxy_protocol.erl]: 😁

File: src/ranch_proxy_protocol.erl [hunk 3]
```diff
@@ -260,11 +329,91 @@ parse_proxy_protocol(<<"TCP", Proto:1/binary, _:1/binary, Info/binary>>) ->
...
+    netns;
+pp2_type(_) ->
+    invalid_pp2_type.
+
+pp2_value(?PP2_TYPE_SSL, <<Client:1/binary, _:32, Rest/binary>>) ->
+    case pp2_client(Client) of
+        invalid_client ->
+            invalid;
+        _ ->
+            %% Can validate we get all necessary values based on the Client bit field here
```
Comment (evanmcc) [inline src/ranch_proxy_protocol.erl]: huh?

**Comment by ferd** [comment]
@evanmcc fixed

**Comment by evanmcc** [comment]
+1

**Pull request merged**